adapters/lockerdome/lockerdome.go

```diff
@@ -0,0 +1,152 @@
+package lockerdome
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const unexpectedStatusCodeMessage = "Unexpected status code: %d. Run with request.debug = 1 for more info"
+
+// Implements Bidder interface.
+type LockerDomeAdapter struct {
+	endpoint string
+}
+
+// MakeRequests makes the HTTP requests which should be made to fetch bids [from the bidder, in this case, LockerDome]
+func (adapter *LockerDomeAdapter) MakeRequests(openRTBRequest *openrtb.BidRequest, extraReqInfo *adapters.ExtraRequestInfo) (requestsToBidder []*adapters.RequestData, errs []error) {
+
+	numberOfImps := len(openRTBRequest.Imp)
+
+	if openRTBRequest.Imp == nil || numberOfImps == 0 { // lockerdometest/supplemental/empty_imps.json
+		err := &errortypes.BadInput{
+			Message: "No valid impressions in the bid request.",
+		}
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	var indexesOfValidImps []int
+	for i := 0; i < numberOfImps; i++ {
+		// LockerDome currently only supports banner impressions, and requires data in the ext field.
+		if openRTBRequest.Imp[i].Banner == nil { // lockerdometest/supplemental/unsupported_imp_type.json
+			err := &errortypes.BadInput{
+				Message: "LockerDome does not currently support non-banner types.",
+			}
+			errs = append(errs, err)
+			continue
+		}
+		var bidderExt adapters.ExtImpBidder
+		err := json.Unmarshal(openRTBRequest.Imp[i].Ext, &bidderExt)
+		if err != nil { // lockerdometest/supplemental/no_ext.json
+			err = &errortypes.BadInput{
+				Message: "ext was not provided.",
+			}
+			errs = append(errs, err)
+			continue
+		}
+		var lockerdomeExt openrtb_ext.ExtImpLockerDome
+		err = json.Unmarshal(bidderExt.Bidder, &lockerdomeExt)
+		if err != nil { // lockerdometest/supplemental/no_adUnitId_param.json
+			err = &errortypes.BadInput{
+				Message: "ext.bidder.adUnitId was not provided.",
+			}
+			errs = append(errs, err)
+			continue
+		}
+		if lockerdomeExt.AdUnitId == "" { // lockerdometest/supplemental/empty_adUnitId_param.json
+			err := &errortypes.BadInput{
+				Message: "ext.bidder.adUnitId is empty.",
+			}
+			errs = append(errs, err)
+			continue
+		}
+		indexesOfValidImps = append(indexesOfValidImps, i)
+	}
+
+	var validImps []openrtb.Imp
```

Comment:
One last comment @margsliu this is small but can improve efficiency. Do you think adding an `if` statement in line 72 makes sense? In other words, filter out the "bad" imps only if there are "bad" imps to discard.
```
 22 func (adapter *LockerDomeAdapter) MakeRequests(openRTBRequest *openrtb.BidRequest, extraReqInfo *adapters.ExtraRequestInfo) (requestsToBidder []*adapters.RequestData, errs []error>
 23
 24     numberOfImps := len(openRTBRequest.Imp)
 25
 26 +--  7 lines: if openRTBRequest.Imp == nil || numberOfImps == 0 { lockerdometest/supplemental/empty_imps.json-----------------------------------------------------------------------
 33
 34     var indexesOfValidImps []int
 35     for i := 0; i < numberOfImps; i++ {
 36         // LockerDome currently only supports banner impressions, and requires data in the ext field.
 37 +-- 25 lines: if openRTBRequest.Imp[i].Banner == nil { lockerdometest/supplemental/unsupported_imp_type.json------------------------------------------------------------------------
 62         if lockerdomeExt.AdUnitId == "" { // lockerdometest/supplemental/empty_adUnitId_param.json
 63             err := &errortypes.BadInput{
 64                 Message: "ext.bidder.adUnitId is empty.",
 65             }
 66             errs = append(errs, err)
 67             continue
 68         }
 69         indexesOfValidImps = append(indexesOfValidImps, i)
 70     }
 71
 72  +  if numberOfImps > len(indexesOfValidImps) {
 73         var validImps []openrtb.Imp
 74         for j := 0; j < len(indexesOfValidImps); j++ {
 75             validImps = append(validImps, openRTBRequest.Imp[j])
 76         }
 77         if len(validImps) == 0 {
 78             err := &errortypes.BadInput{
 79                 Message: "No valid or supported impressions in the bid request.",
 80             }
 81             errs = append(errs, err)
 82             return nil, errs
 83         } else {
 84             openRTBRequest.Imp = validImps
 85         }
 86  +  }
 87     openRTBRequestJSON, err := json.Marshal(openRTBRequest)
 88     if err != nil {
 89         errs = append(errs, err)
 90         return nil, errs
 91     }
 92
 93     headers := http.Header{}
 94 +-- 12 lines: headers.Add("Content-Type", "application/json;charset=utf-8")---------------------------------------------------------------------------------------------------------
106
107     return requestsToBidder, nil
108
109 }
adapters/lockerdome/lockerdome.go [+]
```

Reply:
Yeah! Sounds good. I've added the `if`. Thank you :)
Test coverage is at 94% now.